### PR TITLE
Created AbstractAuthenticator and made necessary updates to LdapLogin

### PR DESCRIPTION
**Pull request by marcus-oscarsson**
**Greetings from the MXCuBE meeting :)**

Introduced `AbstractAuthenticator` and made necessary updates to `LdapLogin`

This also means that we should move the `authenticate` call from `ISPyBclient` to the the authenticate/login
logic. 

All `*LdapLogin` classes have been updated so that they care now call `*Authenticator`

### CODE REVIEW

**Comment by marcus-oscarsson** [comment]
@rhfogh, @MartinSavko: Do you see an issue with this on the Qt client side ?

**Comment by rhfogh** [comment]
I am a little confused. I do not see AbstractAuthenticator or the relevant mockup anywhere in either code or config files, in the PR or the pre-existing code. Should they not be there?

**Review by rhfogh — CHANGES_REQUESTED** [review]
Please clarify where the AnstractAuthernticator mentioned is supposed to live.

**Comment by marcus-oscarsson** [comment]
@rhfogh :) thanks, I forgot to add those files, done now

File: mxcubecore/HardwareObjects/LdapAuthenticator.py [hunk 1]
```diff
@@ -72,111 +74,112 @@ def connect(self):
...
+            self._reconnect()
 
         return (False, msg)
 
-    # Check password in LDAP
-    def login(self, username, password, retry=True, fields=None):
+    def get_field_values(self):
+        return self._field_values
+
+    def invlidate(self):
```
Comment (rhfogh) [inline mxcubecore/HardwareObjects/LdapAuthenticator.py]: Typo: invalidate

**Review by rhfogh — CHANGES_REQUESTED** [review]
Fix typos. Apart from that OK on my side - a standard mockup run works without trouble. But then I do not use authentication anyway, so I can hardly speak for that part.

File: mxcubecore/HardwareObjects/LdapAuthenticator.py (hunk 1 above)
Comment (marcus-oscarsson) [reply mxcubecore/HardwareObjects/LdapAuthenticator.py]: ;) Thanks !

File: mxcubecore/HardwareObjects/abstract/AbstractAuthenticator.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+
+class AbstractAuthenticator(HardwareObject):
+    __metaclass__ = abc.ABCMeta
+
+    def init(self) -> None:
+        super().init()
+
+    @abc.abstractmethod
+    def authenticte(self, username: str, password: str) -> bool:
```
Comment (rhfogh) [inline mxcubecore/HardwareObjects/abstract/AbstractAuthenticator.py]: Typo: authenticate

This one is still spelled wrong

Comment (marcus-oscarsson) [reply mxcubecore/HardwareObjects/abstract/AbstractAuthenticator.py]: Ohh my, I'm need to go the eye doctor :) thanks !

**Review by rhfogh — APPROVED** [review]

**Comment by fabcor-maxiv** [comment]
Does this pull request imply some changes that a facility should make in their facility-specific code? Or is that intended to be "backwards compatible" so to say?

And this pull request from the web part https://github.com/mxcube/mxcubeweb/pull/1140 is required, is that right? Or are the 2 pull requests independent?

**Comment by marcus-oscarsson** [comment]
This PR and https://github.com/mxcube/mxcubeweb/pull/1140 are independent. It makes it possible to decouple the authentication bit from the ISPyB client (for instance in your case ISPyBrestClient) from the rest. 

*EDIT* The actual decoupling of ISPyB and authentication is for a future PR

File: mxcubecore/HardwareObjects/ISPyBClient.py [hunk 3]
```diff
@@ -739,11 +740,16 @@ def login(self, loginID, psd, ldap_connection=None, create_session=True):
         }
 
     def ldap_login(self, login_name, psd, ldap_connection):
+        warnings.warn(
+            ("Using Authenticatior from ISPyBClient is depricated,
+            "use Authenticator to authenticate spereatly and then login to ISPyB"),
```
Comment (fabcor-maxiv) [inline mxcubecore/HardwareObjects/ISPyBClient.py:745]: This is not valid syntax, is it? Maybe missing a closing `"` on line 744.
I am confused, if I am right and it is not valid syntax, then shouldn't it immediately fail? :D

(Also typo: "deprecated" not "depricated".)

Comment (rhfogh) [reply mxcubecore/HardwareObjects/ISPyBClient.py:745]: Indeed - this is invalid - my editor says so too. Maybe ISPyBClient is not imported?

Comment (marcus-oscarsson) [reply mxcubecore/HardwareObjects/ISPyBClient.py:745]: Yes, it was missing a "

**Pull request merged**